added nonhydrostatic capabilities to MPAS-O

scalandr opened the pull request:
This is a nonhydrostatic version of MPAS-O. This nonhydrostatic model enables new physics, allowing the correct representation of, for example, high-resolution internal waves dynamics, which are not well represented by the hydrostatic approximation. New vertical momentum and pressure terms are added in the formulation, and a coupled elliptic system is formed for the solution of the nonhydrostatic pressure correction. At every time step, this system is solved with the PETSc library using preconditioned iterative Krylov solvers. The flag `config_enable_nonhydrostatic_mode` can be used to turn on and off the nonhydrostatic capabilities.

xylar commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -1460,6 +1595,7 @@
 		<package name="tidalPotentialForcingPKG" description="This package controls variables required for tidal potential forcing"/>
 		<package name="topographicWaveDragPKG" description="This package controls variables required for topographic wave drag"/>
 		<package name="gotmPKG" description="This package is for GOTM variables, which are only needed when using GOTM for the vertical turbulence closure."/>
+                <package name="nonHydrostaticMode" description="This package is for variables needed for the nonHydrostatic capabilities"/>
```
```suggestion
                <package name="nonhydrostaticPKG" description="This package is for variables needed for the nonHydrostatic capabilities"/>
```
All packages should end in `PKG` for clarity.  I think we want to always use `nonhydrostatic`, not `nonHydrostatic`.

xylar commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -2068,6 +2204,18 @@
 		<var name="ssh" type="real" dimensions="nCells Time" units="m"
 			 description="sea surface height"
 		/>
+                <var name="verticalVelocity" type="real" dimensions="nVertLevelsP1 nCells Time" units="m s^{-1}"
+       		         description="prognostic vertical velocity, only used when config_enable_nonhydrostatic_mode = .true."
+    		/>
```
There is already a `vertVelocityTop` that seems identical to this except that it is not prognostic.  
https://github.com/E3SM-Ocean-Discussion/E3SM/blob/6d69143ba38a5592c1be2097bd45b8e4a60a75a8/components/mpas-ocean/src/Registry.xml#L2794-L2797
It seems confusing to have both.  I wonder what we can do about this.

scalandr replied:
@vanroekel, based on Xylar's comment, what do you think if we change the name from `verticalVelocity` to `vertVelocityNonhydro`?

scalandr replied:
@xylar, I've talked with Luke and Darren, what do you think if we change the name from `verticalVelocity` to `vertVelocityNonhydro`?

xylar replied:
Sure, I think that works.

xylar commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -2068,6 +2204,18 @@
 		<var name="ssh" type="real" dimensions="nCells Time" units="m"
 			 description="sea surface height"
 		/>
+                <var name="verticalVelocity" type="real" dimensions="nVertLevelsP1 nCells Time" units="m s^{-1}"
+       		         description="prognostic vertical velocity, only used when config_enable_nonhydrostatic_mode = .true."
+    		/>
+                <var name="nonhydrostaticPressure" type="real" dimensions="nVertLevels nCells Time" units="N m^{-2}"
+       	                 description="nonHydrostatic pressure , set to zero if config_enable_nonhydrostatic_mode = .false."
+                />
+                <var name="tke" type="real" dimensions="nVertLevels nCells Time" units="m^2 s^{-2}"
+                         description="Turbulence Kinetic Energy for the two equation turbulence closure"
+                />
+                <var name="glsPsi" type="real" dimensions="nVertLevels nCells Time" units="variable"
+                         description="psi variable for the two equation turbulence closure (see Umlauf and Burchard 2003).  The units here depend on the choice of variables in the GLS scheme"
+                />
```
Here and below in the registry, it would be best not to introduce new variables that aren't used when not in nonhydro mode because of the extra memory this incurs.  If at all possible, these should all be in the `nonhydrostaticPKG`.  So far, you have created the package but not added any variables to it.

xylar commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -2299,6 +2447,15 @@
 		<var name="bed_elevation" type="real" dimensions="nCells" units="m"
 			 description="Depth of the bottom of the ocean. Given as negative distance from sea level. Used as input to replace topography input file."
 		/>
+                <var name="boundaryCellMask" type="integer" dimensions="nVertLevels nCells" units="unitless"
+                         description="cell mask for boundary cells"
+                />
+    		<var name="boundaryNormalAngle" type="real" dimensions="nVertLevels nCells" units="radians"
+       			 description="angle of the boundary normal relative to east/west"
+    		/>
+                <var name="distanceToBoundary" type="real" dimensions="nVertLevels nCells" units="m"
+                         description="distance to solid boundary, only used for idealized case"
+                />
```
Add to `nonhydrostaticPKG`?

xylar commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -2411,6 +2568,9 @@
 			 description="time tendency of normal component of velocity"
 			 packages="forwardMode"
 		/>
+                <var name="tendVerticalVelocity" type="real" dimensions="nVertLevelsP1 nCells Time" units="m s^{-2}" name_in_code="verticalVelocityTend"
+                         description="time tendency of the vertical velocity (Only active for non hydrostatic mode)"
+                />
```
Add to` nonhydrostaticPKG`?

I think having the name be `tendVerticalVelocity` but `name_in_code` be `verticalVelocityTend` is too confusing.  For consistency with other variables in this pool, the `name_in_code` should be `verticalVelocity`.

xylar commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -2419,6 +2579,13 @@
 			 description="time tendency of sea-surface height"
 			 packages="forwardMode"
 		/>
+                <var name="tendTKE" type="real" dimensions="nVertLevels nCells Time" units="m^2 s^{-3}" name_in_code="tendTKE"
+                         description="time tendency of turbulence kinetic energy"
+                         packages="forwardMode"
+                />
+                <var name="tendglsPsi" type="real" dimensions="nVertLevels nCells Time" units="m^2 s^{-4}" name_in_code="tendglsPsi"
+                         description="time tendency of psi variable from Umlauf and Burchard (2003)"
+                />
```
Same as above.  Either the `name_in_code` should be `tke` and `gslPsi` for consistency.

xylar replied:
Add to `nonhydrostaticPKG`

xylar commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -2419,6 +2579,13 @@
+                         description="time tendency of turbulence kinetic energy"
+                         packages="forwardMode"
+                />
+                <var name="tendglsPsi" type="real" dimensions="nVertLevels nCells Time" units="m^2 s^{-4}" name_in_code="tendglsPsi"
+                         description="time tendency of psi variable from Umlauf and Burchard (2003)"
+                />
```
Should the name be `tendGlsPsi` or `tendGLSPsi` for easier parsing?

xylar commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -2438,6 +2605,9 @@
 		<var name="daysSinceStartOfSim" type="real" dimensions="Time" units="days"
 			 description="Time since simulationStartTime, for plotting"
 	 	/>
+                <var name="nhPressureCorrection" type="real" dimensions="nVertLevels nCells Time" units="unitless"
+                         description="the pressure correction coming from the poisson pressure equation"
+                />
```
Add to `nonhydrostaticPKG`

xylar commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -2453,6 +2623,24 @@
 		<var name="temperatureShortWaveTendency" dimensions="nVertLevels nCells Time" units="degrees Celsius per second" packages="tracerBudget" name_in_code="temperatureShortWaveTendency" type="real"
 				 description="potential temperature tendency due to penetrating shortwave"
 		/>
+                <var name="velocityHmixTend" dimensions="nVertLevels nEdges Time" units="m/s^2" type="real"
+                  description="tendency of normal velocitydue to horizontal mixing"
+                />
+                <var name="velocityVmixTend" dimensions="nVertLevels nEdges Time" units="m/s^2" type="real"
+                  description="tendency of normal velocity due to vertical mixing"
+                />
+                <var name="velocityCoriolisTend" dimensions="nVertLevels nEdges Time" units="m/s^2" type="real"
+                  description="tendency of normal velocity due to Coriolis"
+                />
+                <var name="velocityPGFtend" dimensions="nVertLevels nEdges Time" units="m/s^2" type="real"
+                  description="tendency of normal velocity due to PGF"
+                />
+                <var name="velocityVadvTend" dimensions="nVertLevels nEdges Time" units="m/s^2" type="real"
+                  description="tendency of normal velocity due to vertical advection"
+                />
+                <var name="velocityForcingTend" dimensions="nVertLevels nEdges Time" units="m/s^2" type="real"
+                  description="tendency of normal velocity due to top and bottom forcing"
+                />
```
Add to `nonhydrostaticPKG`

xylar replied:
I believe this still needs to happen.

xylar commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -2806,6 +2994,18 @@
 			 description="vertical viscosity defined at the cell center (horizontally) and top (vertically)"
 			 packages="forwardMode;analysisMode"
 		/>
+                <var name="dynamicViscosity" type="real" dimensions="nVertLevelsP1 nCells Time" units="Ns/m^2"
+                         description="dynamic viscosity of water as a function of temperature and salinity"
+                         packages="forwardMode;analysisMode"
+                />
+                <var name="uStar" type="real" dimensions="nVertLevels nCells Time" units="m s^{-1}"
+                  	 description="stress velocity for wall bounded flows"
+                  	 packages="forwardMode;analysisMode"
+    		/>
+    		<var name="yPlus" type="real" dimensions="nVertLevels nCells Time" units="m"
+       			 description="boundary normal coordinate"
+      		         packages="forwardMode;analysisMode"
+    		/>
```
Add to `nonhydrostaticPKG`

xylar commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -2867,6 +3067,27 @@
 		<var name="indMLD" type="integer" dimensions="nCells Time" units="unitless"
 			description="index of model where mixed layer depth occurs (always one past)"
 		/>
+                <var name="ugrad" type="real" dimensions="R3 nVertLevels nCells Time" units="s^{-1}"
+                        description="three-dimensional gradient of zonal velocity"
+                />
+                <var name="vgrad" type="real" dimensions="R3 nVertLevels nCells Time" units="s^{-1}"
+                        description="three-dimensional gradient of meridional velocity"
+                />
+                <var name="wgrad" type="real" dimensions="R3 nVertLevels nCells Time" units="s^{-1}"
+                        description="three-dimensional gradient of vertical velocity"
+                />
+                <var name="tgrad" type="real" dimensions="R3 nVertLevels nCells Time" units="s^{-1}"
+                        description="three-dimensional gradient of temperature"
+                />
+                <var name="sgrad" type="real" dimensions="R3 nVertLevels nCells Time" units="s^{-1}"
+                        description="three-dimensional gradient of salinity"
+                />
+                <var name="Sh" type="real" dimensions="nVertLevels nCells Time" units="unitless"
+                        description="Structure function for heat and salt -- may want to consider separating at some point"
+                />
+                <var name="Sm" type="real" dimensions="nVertLevels nCells Time" units="unitless"
+                        description="Structure function for momentum"
+                />
```
Add to `nonhydrostaticPKG`

xylar replied:
I'm worried that the names `Sh` and `Sm` are not sufficiently descriptive or unique.

xylar commented on components/mpas-ocean/src/shared/Makefile:
```diff
@@ -27,6 +27,13 @@ OBJS = mpas_ocn_init_routines.o \
 	   mpas_ocn_vel_forcing_surface_stress.o \
 	   mpas_ocn_vel_forcing_explicit_bottom_drag.o \
 	   mpas_ocn_vel_pressure_grad.o \
+		 mpas_ocn_vvel_vert_advection.o \
+		 mpas_ocn_vvel_horiz_advection.o \
+		 mpas_ocn_vvel_coriolis.o \
+		 mpas_ocn_vvel_hmix_del2.o \
+		 mpas_ocn_vvel_advection.o \
+		 mpas_ocn_vvel_pgf.o \
+		 mpas_ocn_nonhydrostatic_pressure_solve.o \
```
The indentation here seems to be incorrect.

xylar commented on components/mpas-ocean/src/shared/Makefile:
```diff
@@ -74,12 +81,20 @@ OBJS = mpas_ocn_init_routines.o \
 	   mpas_ocn_vel_forcing_topographic_wave_drag.o \
 	   mpas_ocn_transport_tests.o \
 	   mpas_ocn_vel_self_attraction_loading.o 
+           mpas_ocn_two_equation_model_driver.o \
+           mpas_ocn_two_equation_structure_functions.o \
+           mpas_ocn_two_equation_model_driver.o \
+           mpas_ocn_tke_glsPsi_transport.o \
+           mpas_ocn_tke_advection_mono.o \
+           mpas_ocn_glsPsi_advection_mono.o \
+           mpas_ocn_shear_production.o \
+           mpas_ocn_buoyancy_production.o
```
Indentation here needs to be with tabs, not spaces.  As a result, I am not able to build the code on Anvil.

xylar reviewed (changes requested):
@scalandr, thank you for this amazing work!  This will be such an great new capability for us to have.

I have noticed several things that need a little work. I still have a lot of code to go through but this is all I have time for for a little while so wanted to give you a chance to address it first.

The major one is that most if not all of the variables you added should be in the `nonhydrostaticPKG` package (I suggest this name change).  In the code, it will be important to only access these fields if the package is active.  Typically, you can check if the field pointers are associated.  See many examples of this throughout the code.  If you don't make use of packages, your changes here will add substantially to the memory usage of the model even in hydrostatic mode.

The other, which is preventing the code from compiling and therefore preventing me from verifying that the results in hydrostatic mode are bit-for-bit is the incorrect indentation in `src/shared/Makefile`.  It needs to be with tabs, not spaces.

Please verify that the code compiles for you and then ping me to try testing again.

scalandr commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -424,6 +432,75 @@
 					possible_values="Any positive real value."
 		/>
 	</nml_record>
+        <nml_record name="twoEquationModel" mode="forward">
```
@xylar, I talked with Luke about this and we decided to eliminate the `two_equation_model` from this PR request. That part of the code is still under development and it is not needed to run the classical nonhydro test cases (the ones in the nonhydro test group).

scalandr replied:
Got rid of two equation model.

scalandr commented on components/mpas-framework/Makefile:
```diff
@@ -926,7 +945,7 @@ pio_test:
 	@#
 	@echo "Checking for a usable PIO library..."
 	@($(FC) pio1.f90 $(FCINCLUDES) $(FFLAGS) $(LDFLAGS) $(LIBS) -o pio1.out &> /dev/null && echo "=> PIO 1 detected") || \
-	 ($(FC) pio2.f90 $(FCINCLUDES) $(FFLAGS) $(LDFLAGS) $(LIBS) -o pio2.out &> /dev/null && echo "=> PIO 2 detected") || \
+	 ($(FC) pio2.f90 $(FCINCLUDES) $(FFLAGS) $(LDFLAGS) $(LIBS) -o pio2.out && echo "=> PIO 2 detected") || \
```
@xylar, done.

scalandr commented on components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_rk4.F:
```diff
@@ -495,27 +561,49 @@ subroutine ocn_time_integrator_rk4(domain, dt)!{{{
              call mpas_timer_stop("RK4-prevent drying")
         end if
 
+        !FIXME - LVR - ALE is not working with non hydrostatic, for now implement switch to disable
```
Done.

scalandr commented on components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_rk4.F:
```diff
@@ -495,27 +561,49 @@ subroutine ocn_time_integrator_rk4(domain, dt)!{{{
         ! In RK4 notation, we are computing the right hand side f(t,y),
         ! which is the same as k_j / h.
         call mpas_timer_start("RK4 vel/thick tendency computations")
-
```
Done.

scalandr commented on components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_rk4.F:
```diff
@@ -601,15 +689,21 @@ subroutine ocn_time_integrator_rk4(domain, dt)!{{{
       !  A little clean up at the end: rescale tracer fields and compute diagnostics for new state
       !
       call mpas_timer_start("RK4-cleanup phase")
-
```
Done.

scalandr commented:
Hi @xylar, I am ready to ask for a re-review of the PR. Before I click on the "re-request review" button I wanted to ask you about the two conflicting files. If I click on "Resolve conflicts" and fix them there, then my local branch will not see the changes, correct? What is your recommendation on how to proceed? Thanks!

xylar commented:
@scalandr, my recommendation for resolving conflicts like this is always to rebase onto `E3SM-Project/E3SM/master`.

If you use "Resolve conflicts" here, it will do a merge of `E3SM-Ocean-Discussion/E3SM/master` to this branch, but that's not very helpful since that isn't the same thing as `E3SM-Project/E3SM/master` and some merge conflicts might remain.  Also, I am a strong advocate of avoiding merging at nearly all cost expect for merging pull requests.  If you merge `E3SM-Project/E3SM/master` into your branch instead of rebasing, things can become very messy to untangle later.  (Of course, if mistakes happen during a rebase, that can also be tricky to untangle later, but I still very much think it's worth it.)

I will update `E3SM-Ocean-Discussion/E3SM/master` to match `E3SM-Project/E3SM/master` right now.  That way, if you rebase onto `E3SM-Project/E3SM/master`, you won't see any extra commits that aren't yours.

xylar commented:
Okay, I just updated `E3SM-Ocean-Discussion/E3SM/master`.  This shows why it's important to use `E3SM-Project/E3SM/master` as the base branch instead: there are now 4 conflicting files, not jut 2.

scalandr commented:
@xylar, when we did the rebase together last time we did
```
git remote add E3SM-Project/E3SM [git@github.com:E3SM-Project/E3SM.git](mailto:git@github.com:E3SM-Project/E3SM.git)
git fetch --all -p
git rebase -i E3SM-Project/E3SM/master
```
Shall i just do `git fetch --all -p` before the `git rebase` command this time?

xylar commented:
@scalandr, yes, that's right.  You don't need to add `E3SM-Porject/E3SM` as a remote since it's already been added.

xylar commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -2118,6 +2185,12 @@
 		<var name="ssh" type="real" dimensions="nCells Time" units="m"
 			 description="sea surface height"
 		/>
+                <var name="vertVelocityNonhydro" type="real" dimensions="nVertLevelsP1 nCells Time" units="m s^{-1}"
+			 description="prognostic vertical velocity, only used when config_enable_nonhydrostatic_mode = .true."
+    		/>
+                <var name="nonhydrostaticPressure" type="real" dimensions="nVertLevels nCells Time" units="N m^{-2}"
+			 description="nonhydrostatic pressure, set to zero if config_enable_nonhydrostatic_mode = .false."
+                />
```
Can these variables be put in the `nonhydrostaticPKG` package?  If they aren't used at all when the `nonhydrostaticPKG` is disabled, this should be possible.  If not, this will add to the memory used in hydrostatic mode.

xylar commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -2466,6 +2539,9 @@
 			 description="time tendency of normal component of velocity"
 			 packages="forwardMode"
 		/>
+                <var name="tendVertVelocityNonhydro" type="real" dimensions="nVertLevelsP1 nCells Time" units="m s^{-2}" name_in_code="vertVelocityNonhydroTend"
+                         description="time tendency of the vertical velocity (Only active for non hydrostatic mode)"
+                />
```
Same here, can this be in the `nonhydrostaticPKG` package?

xylar commented on components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_rk4.F:
```diff
@@ -251,6 +265,8 @@ subroutine ocn_time_integrator_rk4(domain, dt)!{{{
          call mpas_pool_get_array(meshPool, 'maxLevelCell', maxLevelCell)
          call mpas_pool_get_array(meshPool, 'maxLevelEdgeTop', maxLevelEdgeTop)
 
+         print*, dt
+
```
Please take this out.  I assume it's for debugging.

xylar commented on components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_rk4.F:
```diff
@@ -503,19 +531,26 @@ subroutine ocn_time_integrator_rk4(domain, dt)!{{{
         block => domain % blocklist
         do while (associated(block))
            call ocn_time_integrator_rk4_compute_vel_tends(domain, block, dt, rk_substep_weights(rk_step), domain % dminfo, err )
-
            call ocn_time_integrator_rk4_compute_thick_tends( block, dt, rk_substep_weights(rk_step), err )
            block => block % next
         end do
-
         call mpas_timer_stop("RK4 vel/thick tendency computations")
```
Maybe put back these blank lines?

xylar commented on components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_rk4.F:
```diff
@@ -503,19 +531,26 @@ subroutine ocn_time_integrator_rk4(domain, dt)!{{{
-
         call mpas_timer_stop("RK4 vel/thick tendency computations")
 
-        ! Update halos for prognostic variables.
```
Maybe put back this comment (below the if)

xylar commented on components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_rk4.F:
```diff
@@ -659,6 +696,21 @@ subroutine ocn_time_integrator_rk4(domain, dt)!{{{
+         call mpas_dmpar_field_halo_exch(domain, 'nhPressureCorrection')
+         call mpas_timer_stop('RK4 NH pressure tend halo pressure')
+
+            call ocn_nonhydrostatic_pressure_update_velocity(normalVelocityNew, vertVelocityNonhydroNew, &
+                 layerThicknessNew, layerThickEdgeMean, nonhydrostaticPressure, nonhydrostaticPressureOld, &
+                 nhPressureCorrection, dt)
```
This is indented 3 spaces too many

xylar commented on components/mpas-ocean/src/shared/mpas_ocn_mesh.F:
```diff
@@ -658,7 +658,7 @@ subroutine ocn_meshCreate(domain) !{{{
          !$acc                   localVerticalUnitVectors, &
          !$acc                   cellTangentPlane,         &
-         !$acc                   coeffs_reconstruct)
+         !$acc                   coeffs_reconstruct)       
```
trailing space should be removed.

xylar reviewed (commented):
I see a few basically cosmetic things that I'd like to see fixed.  After that, I'm ready for this to move to E3SM.

I'd like to have @dengwirda and @mark-petersen both approve here first, though.

scalandr commented:
Hi @xylar, I did a second rebase and took care of all your comments except one regarding `velocityHmixTend`, `velocityVmixTend`, `velocityCoriolisTend`, ect. Those tendencies are computed in both hydro and nonhydro mode, therefore I did not put them in the nonhydro package. I asked @dengwirda and @vanroekel to review the PR, so that we can move forward to E3SM (once it gets approved).

xylar commented:
>  I did a second rebase and took care of all your comments except one regarding velocityHmixTend, velocityVmixTend, velocityCoriolisTend, ect. Those tendencies are computed in both hydro and nonhydro mode, therefore I did not put them in the nonhydro package. 

Oh, okay.  Could you explain a bit more about those?  I assumed they would be needed only for nonhydro because they are added in this PR.  If they are also needed by hydro, it would be important to understand why new variables are needed for that mode and whether there is any risk for non-bit-for-bit changes in hydro as a result.

xylar commented:
Looking briefly through the code again, it seems like `velocityHmixTend`, etc. are computed but not used in any other computation so they are probably just for diagnostic purposes (e.g. for energy or momentum budgets).  I'm wondering if it might make sense to introduce those variables in a separate PR that could be reviewed on its own, since this PR is so giant.  @vanroekel and @dengwirda, what do you think?

vanroekel commented:
I agree with @xylar about the `velTends`. I added those solely as diagnostics.  Removing those from this PR makes a lot of sense to me.  I’m not convinced they are actually useful to have given how hard it is to interpret normal velocity.  But if included they should be their own PR.

xylar reviewed (approved):
Thanks @scalandr!  I'm ready for this to move to E3SM as soon as @dengwirda and @vanroekel get a chance to approve here.

vanroekel commented on components/mosart/src/riverroute/RtmMod.F90:
```diff
@@ -4212,7 +4215,7 @@ subroutine MOSART_init
 
   ! retrieve the downstream channel attributes after some post-processing above
-  if (Tctl%RoutingMethod == 2 ) then       ! Use diffusion wave method in channel routing computation.
+  if (Tctl%RoutingMethod == DW ) then       ! Use diffusion wave method in channel routing computation.
```
Are these mosart changes due to differences between the head of ocean discussions and e3sm/master?

xylar replied:
They shouldn't be.  I just updated E3SM-Ocean-Discussion `master` to match E3SM-Project's `master` and they're still here.

There are 7 odd commits listed in this PR that shouldn't be here.  @scalandr, if you can rebase once again onto E3SM-Project/E3SM/master (i.e. `origin/master`) and remove the first 7 commits at the top from the list, that should take care of it.  It should be an easy rebase because no conflicts are listed here on GitHub.

vanroekel commented on components/mpas-ocean/bld/build-namelist:
```diff
@@ -797,6 +797,21 @@ add_default($nl, 'config_topographic_wave_drag_coeff');
+add_default($nl, 'config_petsc_atol');
+add_default($nl, 'config_petsc_maxit');
+add_default($nl, 'config_use_constant_forced_pgf');
+
```
were these added by the automated scripts?  Or hand entered?

vanroekel commented on components/mpas-ocean/bld/namelist_files/namelist_defaults_mpaso.xml:
```diff
@@ -365,6 +365,18 @@
+<config_petsc_rtol>1.0e-10</config_petsc_rtol>
+<config_petsc_atol>1.0E-20</config_petsc_atol>
+<config_petsc_maxit>10000</config_petsc_maxit>
+<config_use_constant_forced_pgf>.false.</config_use_constant_forced_pgf>
```
I'm not sure we want to retain this option long term.  My feeling is to remove the constant_forced_pgf from this PR

vanroekel commented on components/mpas-ocean/bld/namelist_files/namelist_definition_mpaso.xml:
```diff
@@ -1732,6 +1732,88 @@ Valid values: any positive real, but typically 1000-1035
+
+<entry id="config_use_constant_forced_pgf" type="logical"
+        category="nonhydrostatic_pressure" group="nonhydrostatic_pressure">
+logical to enable a constant pressure gradient that is added to rest of PGF NOTE: need to remove after rebasing on to master...
```
Here is actually a note (I'm guessing I made it) to remove this flag.

vanroekel commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -1305,6 +1355,10 @@
 		/>
+                <nml_option name="config_disable_ALE" type="logical" default_value=".false." units="unitless"
+                                        description="flag to disable the ALE coordinate and revert to eulerian mode NOTE: only works for RK4 at present"
+                                        possible_values=".true. or .false."
```
This is another really untested option that I would suggest we remove.  I don't think we want this finding its way to e3sm master

vanroekel commented on components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_rk4.F:
```diff
@@ -153,18 +156,22 @@ subroutine ocn_time_integrator_rk4(domain, dt)!{{{
       real (kind=RKIND), dimension(:,:), pointer :: layerThicknessCur, layerThicknessNew
       real (kind=RKIND), dimension(:,:), pointer :: highFreqThicknessCur, highFreqThicknessNew
       real (kind=RKIND), dimension(:,:), pointer :: lowFreqDivergenceCur, lowFreqDivergenceNew
+      real (kind=RKIND), dimension(:,:), pointer :: nonhydrostaticPressure, nonhydrostaticPressureOld
```
should these be changed to nonhydrostaticPressureNew, and nonhydrostaticPressureCur for consistency with other variables?

vanroekel commented on components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_rk4.F:
```diff
@@ -1645,6 +1831,11 @@ subroutine ocn_time_integrator_rk4_cleanup(block, dt, err)!{{{
 
       call ocn_vmix_implicit(dt, meshPool, statePool, forcingPool, scratchPool, err, 2)
 
+      if (config_enable_nonhydrostatic_mode .and. .not. config_enable_vertMomentum_disable_nonHydro) then
```
here is another flag to remove -- I don't envision needing config_enable_vertMomentum_disable_nonHydro, but I'm okay with leaving it if we want it for longer term debugging.

vanroekel commented on components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_si.F:
```diff
@@ -270,7 +271,11 @@ subroutine ocn_time_integrator_si(domain, dt)!{{{
+         vertVelocityNonhydroNew,         &! Vertical Velocity at new time
+         vertVelocityNonhydroCur,         &! Vertical Velocity at current time
+         nonhydrostaticPressure,      &! nhpressure at new time
+         nonhydrostaticPressureOld     ! nh pressure at current time
```
same comment as above, should probably rename the nonhydrostaticPressure arrays for consistency with other fields.

vanroekel commented on components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_split.F:
```diff
@@ -216,7 +219,11 @@ subroutine ocn_time_integrator_split(domain, dt)!{{{
+         vertVelocityNonhydroCur,         &! vertical velocity at current time
+         vertVelocityNonhydroNew,         &! vertical velocity at new time
+         nonhydrostaticPressure,      &! nonhydro pressure at new time
+         nonhydrostaticPressureOld     ! nonhdyro pressure at current time
```
again -- suggest renaming these two arrays.

vanroekel commented on components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_split.F:
```diff
@@ -1753,8 +1789,26 @@ subroutine ocn_time_integrator_split(domain, dt)!{{{
+         call mpas_pool_get_array(statePool,   'normalVelocity', &
+                                                normalVelocityCur, 1)
+         call mpas_pool_get_array(tracersTendPool,'activeTracersTend', &
+                                                   activeTracersTend)
```
any ideas why these tracer arrays show up in this PR?  They should be unrelated to any non hydro code

vanroekel commented on components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_split.F:
```diff
@@ -2380,6 +2459,8 @@ subroutine ocn_time_integrator_split(domain, dt)!{{{
 #endif
 
+      call mpas_dmpar_field_halo_exch(domain, 'surfaceFrictionVelocity')
+
```
similarly this should be unrelated to nonhydro, not sure why this is here.

vanroekel commented on components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:
```diff
@@ -1112,7 +1114,7 @@ subroutine ocn_diagnostic_solve_vorticity(dt, normalVelocity, tangentialVelocity
       nCells = nCellsAll
       nVertices = nVerticesAll
-      nEdges = nCellsAll
+      nEdges = nEdgesAll
```
is this a bug fix?  The original definitely seems wrong, but that change is not really a part of the PR ti seems.

scalandr replied:
Yes, this is a bug fix.

vanroekel commented on components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:
```diff
@@ -1688,6 +1691,21 @@ subroutine ocn_diagnostic_solve_vortVel(relativeVorticity, &
+!        !$omp end do
+!        !$omp end parallel
+!      endif
+
```
It's weird to merge in a block of commented out code.  Should we just remove this?  I was never confident this was a correct operation.

xylar replied:
Yes, please remove this.  I agree that we have a general policy against committing commented-out code.

vanroekel commented on components/mpas-ocean/src/shared/mpas_ocn_tendency.F:
```diff
@@ -235,6 +241,133 @@ subroutine ocn_tend_thick(tendPool, forcingPool)!{{{
+        tend_vertVelocityNonhydro, err)
+
+   !
+   ! velocity tendency: del2 dissipation, CALL STRAIGHT to hmix_del2 FOR NOW!!!
```
do we want to change this comment?  The call to hmix_del2 directly is quite different than tracer and normal velocity.  It may be fine to just make the comment read something like 'at present only del2 mixing is implemented for- horizontal mixing of vertical velocity'

vanroekel commented on components/mpas-ocean/src/shared/mpas_ocn_tendency.F:
```diff
@@ -440,7 +578,7 @@ subroutine ocn_tend_vel(domain, tendPool, statePool, forcingPool, &
 
       ! Add horizontal mixing
-      call ocn_vel_hmix_tend(divergence, relativeVorticity, tendVel,err)
+      call ocn_vel_hmix_tend(divergence, relativeVorticity, viscosity, tendVel,err)
```
note that having the 'viscosity' array appear here may be a hold over from the two equation turbulence model code.

vanroekel commented on components/mpas-ocean/src/shared/mpas_ocn_vel_hadv_coriolis.F:
```diff
@@ -90,7 +90,7 @@ subroutine ocn_vel_hadv_coriolis_tend(normRelVortEdge, &
          layerThickEdgeFlux,     &!< [in] layer thickness on edge
          normalVelocity,     &!< [in] Horizontal velocity
-         kineticEnergyCell    !< [in] Kinetic Energy
+         kineticEnergyCell    !< [in] Kinetic Energyi
```
i think there is a typo here "Energyi"

vanroekel commented on components/mpas-ocean/src/shared/mpas_ocn_vel_hmix.F:
```diff
@@ -80,7 +80,7 @@ module ocn_vel_hmix
 !-----------------------------------------------------------------------
 
-   subroutine ocn_vel_hmix_tend(div, relVort, tend, err)!{{{
+   subroutine ocn_vel_hmix_tend(div, relVort, viscosity, tend, err)!{{{
```
I think you should remove viscosity from this routine.

vanroekel commented on components/mpas-ocean/src/shared/mpas_ocn_vel_pressure_grad.F:
```diff
@@ -158,7 +160,9 @@ subroutine ocn_vel_pressure_grad_tend(ssh, pressure, surfacePressure, &
-         SL, SR             ! salinity neighbors for Jacobian
+         SL, SR, dqdza      ! salinity neighbors for Jacobian
+
+      real (kind=RKIND), dimension(:), allocatable :: dqdz
```
I suggest removing this.

vanroekel commented on components/mpas-ocean/src/shared/mpas_ocn_vmix.F:
```diff
@@ -1386,6 +1532,70 @@ subroutine ocn_vmix_init(domain, err)!{{{
+      end do
+
+   end subroutine tridiagonal_solve !}}}
+
```
I don't think this is a good addition.  I believe the performance team has inlined all the tri-diagonal solves and we probably shouldn't add it back.  But maybe since it won't be called in e3sm simulations for some time it's okay as is.  What do you think @philipwjones?

philipwjones replied:
Indeed. You can either use the other routines as a model or look to see if one of the other versions would work with some suitable modifications/pre-computation of input arguments.

vanroekel replied:
@scalandr looks like you got the function inlined.  If so I think you can remove this block of code.

vanroekel commented on components/mpas-seaice/src/shared/mpas_seaice_advection_incremental_remap.F:
```diff
@@ -2394,7 +2394,7 @@ subroutine seaice_run_advection_incremental_remap(&
          dynamicsTimeStep
 
-    logical, pointer :: &
+    logical :: &
```
again - are these diffs due to the state of ocean discussions vs. e3sm/master?

vanroekel reviewed (changes requested):
@scalandr this is a very impressive piece of work.  You've done a superb job streamlining and cleaning this up.  I have a few other suggestions for further cleaning.  Let me know what you think of the possibilities.

scalandr commented:
Hi @vanroekel, I took care of all your comments, so 
- I got rid of unused flags like `config_use_constant_forced_pgf`, `config_disable_ALE`, and `config_enable_vertMomentum_disable_nonHydro`,
I renamed `nonhydrostaticPressure `and `nonhydrostaticPressureOld` to `nonhydrostaticPressureNew` and `nonhydrostaticPressureCur`, respectively, in all time-steppers, 
- I removed `viscosity` from hmix tend, 
- I cleaned other files you mentioned. 
About the vertical mixing in the vertical momentum, I believe I should inline the tri-diagonal solver. It’s something that we’ll have to do eventually, so let’s just do it now. Beside the tri-diagonal solver, the only thing left is the rebase.

scalandr commented:
@xylar, If I do the rebase I see 

<img width="669" alt="rebase" src="https://user-images.githubusercontent.com/10819470/178638346-ec746a22-cc42-4a40-9d49-99d7e4c65106.png">

So should I erase the first 7 lines, i.e. commits 88fb to 7355?

xylar commented:
@scalandr, no, sorry for the confusion.  Here's what we're seeing on GitHub:
![image](https://user-images.githubusercontent.com/4179064/178655549-7a733092-b512-4914-a540-d3d287c85974.png)
Those are the commits that we would want to get rid of.  But I now think it's just GitHub glitching and not noticing that those commits aren't part of this branch.  We've seen that before.  I think if I change the target branch to something else and then change it back to `E3SM-Ocean-Discussion:master`, it should be fine.  I'll try that.

xylar commented:
Indeed, that seems to have taken care of it:
![image](https://user-images.githubusercontent.com/4179064/178655911-e560943e-8f99-41ca-90ec-f861693616b8.png)
@vanroekel, make sure you agree that the commits that shouldn't have been here are gone.

xylar commented:
@scalandr, I noticed when building this today that the make targets are out of date compared to master. Could you do a rebase to bring those changes in when you have time? Not super urgent but it would be a good idea.

scalandr commented:
Hi @xylar, as far as I know @jonbob already rebased this nonhydro branch locally. What would be the best way to proceed? For him to push the rebased version? Thanks!

xylar commented:
@scalandr and @jonbob, that's something for the 2 of you to coordinate but I think that sounds like the easiest solution.  In general, it is very important that no one but the developer force-pushes to a branch on a fork without careful coordination.
